core/touch.js

```diff
@@ -0,0 +1,273 @@
+ * @param {!Event} e Mouse event or touch event.
+ * @return {boolean} Whether the identifier on the event matches the current
+ *     saved identifier.
+ */
+Blockly.Touch.checkTouchIdentifier = function(e) {
+  var identifier;
+  // TODO (fenichel): Improve splitEventByTouches to get rid of this try/catch.
+  try {
+    identifier = (e.changedTouches && e.changedTouches.item(0) &&
+        e.changedTouches.item(0).identifier != undefined &&
+        e.changedTouches.item(0).identifier != null) ?
+        e.changedTouches.item(0).identifier : 'mouse';
+  } catch (ex) {
+    identifier = (e.changedTouches && e.changedTouches[0] &&
+        e.changedTouches[0].identifier != undefined &&
+        e.changedTouches[0].identifier != null) ?
+        e.changedTouches[0].identifier : 'mouse';
+  }
+
+  // if (Blockly.touchIdentifier_ )is insufficient because android touch
+  // identifiers may be zero.
+  if (Blockly.Touch.touchIdentifier_ != undefined &&
+      Blockly.Touch.touchIdentifier_ != null) {
+    // We're already tracking some touch/mouse event.  Is this from the same
+    // source?
+    return Blockly.Touch.touchIdentifier_ == identifier;
+  }
+  if (e.type == 'mousedown' || e.type == 'touchstart') {
+    // No identifier set yet, and this is the start of a drag.  Set it and
+    // return.
+    console.trace('setting touch identfier');
+    Blockly.Touch.touchIdentifier_ = identifier;
+    return true;
+  }
+  // There was no identifier yet, but this wasn't a start event so we're going
+  // to ignore it.  This probably means that another drag finished while this
+  // pointer was down.
+  return false;
+};
+
+/**
+ * Set an event's clientX and clientY from its first changed touch.  Use this to
+ * make a touch event work in a mouse event handler.
+ * @param {!Event} e A touch event.
+ */
+Blockly.Touch.setClientFromTouch = function(e) {
+  if (e.type.indexOf('touch') == 0) {
+    // Map the touch event's properties to the event.
+    var touchPoint = e.changedTouches[0];
+    e.clientX = touchPoint.clientX;
+    e.clientY = touchPoint.clientY;
+  }
+};
+
+/**
+ * Check whether a given event is a mouse or touch event.
+ * @param {!Event} e An event.
+ * @return {boolean} true if it is a mouse or touch event; false otherwise.
+ */
+Blockly.Touch.isMouseOrTouchEvent = function(e) {
+  return e.type.indexOf('touch') == 0 || e.type.indexOf('mouse') == 0;
+};
+
+/**
+ * Split an event into an array of events, one per changed touch or mouse
+ * point.
+ * @param {!Event} e A mouse event or a touch event with one or more changed
+ * touches.
+ * @return {!Array.<!Event>} An array of mouse or touch events.  Each touch
+ *     event will have exactly one changed touch.
+ */
+Blockly.Touch.splitEventByTouches = function(e) {
+  var events = [];
+  if (e.changedTouches) {
+    for (var i = 0; i < e.changedTouches.length; i++) {
+      var newEvent = {
+        type: e.type,
+        changedTouches: [e.changedTouches[i]],
+        target: e.target,
```

Comment:
Interesting.  I might be getting away with this because it's not actually an Event.  As I recall it *does* batter that it has an explicit target, because everything broke down the line otherwise.  And maybe I made it an Object because I ran into the same problem as you did with Events.